eth_utils/abi.py

```diff
@@ -1,68 +1,784 @@
+from collections import (
+    abc,
+)
+import copy
+import itertools
+import re
 from typing import (
     Any,
     Dict,
+    Iterable,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+    cast,
+)
+
+from eth_typing import (
+    ABI,
+    ABIComponent,
+    ABIElement,
+    ABIEvent,
+    ABIFunction,
+)
+
+from eth_utils.types import (
+    is_list_like,
 )
 
 from .crypto import (
     keccak,
 )
 
+T = TypeVar("T")
 
-def collapse_if_tuple(abi: Dict[str, Any]) -> str:
+
+def _align_abi_input(arg_abi: ABIComponent, arg: Any) -> Union[Any, Tuple[Any, ...]]:
+    """
+    Aligns the values of any mapping at any level of nesting in ``arg``
+    according to the layout of the corresponding abi spec.
     """
-    Converts a tuple from a dict to a parenthesized list of its types.
+    tuple_parts = _get_tuple_type_str_and_dims(arg_abi.get("type", ""))
+
+    if tuple_parts is None:
+        # Arg is non-tuple.  Just return value.
+        return arg
+
+    sub_abis: Iterable[ABIComponent] = []
```

Comment:
I get a warning that this line is obsolete since we redefine `sub_abis` in both the `if` block and the `else` block. It'll never be this empty list value.